File: tools/server_observer/observation_session.py
from __future__ import annotations

from copy import deepcopy
from dataclasses import asdict
from dataclasses import dataclass
from dataclasses import field
from pathlib import Path
from time import sleep
from time import time
from typing import Optional

import httpx
from httpx import HTTPTransport

from conflict_interface.data_types.authentication import AuthDetails
from conflict_interface.game_api import GameApi
from conflict_interface.utils.exceptions import AuthenticationException
from tools.server_observer.account import Account
from tools.server_observer.observation_api import ObservationApi
from tools.server_observer.recorder_logger import get_logger
from tools.server_observer.static_map_cache import StaticMapCache
from tools.server_observer.storage import RecordingStorage

logger = get_logger()

MAX_RETRIES = 3
TIME_TILL_RETRY = 10

@dataclass
class ObservationPackage:
    game_id: int = None
    headers: dict = None
    cookies: dict = None
    proxy: dict = None
    auth: AuthDetails = None
    client_version: int = None
    game_server_address: str = None

    time_stamps: dict = field(default_factory=dict)
    state_ids: dict = field(default_factory=dict)

class ObservationSession:
    """
    Holds per-game session context (cookies, headers, proxy, auth) and schedules updates.
    Spawns short-lived ObservationWorker instances to perform individual update requests.
    """

    def __init__(
        self,
            game_id: int,
            account: Account,
            map_cache: StaticMapCache,
            storage_path: Path,
            metadata_path: Optional[Path] = None,
            long_term_storage_path: Optional[Path] = None,
            file_size_threshold: Optional[int] = None
        ):
        self.account = account
        self.map_cache = map_cache
        self.game_id = game_id

        self.storage_path = storage_path
        self.metadata_path = metadata_path
        self.long_term_storage_path = long_term_storage_path
        self.file_size_threshold = file_size_threshold
        self.package = None
        self._shared_transport: Optional[HTTPTransport] = None
        self._storage: Optional[RecordingStorage] = None

        self._start_time: Optional[float] = None
        self._updates_done = 0
        self.next_update_at: float = time()

    def reset(self):
        self.package = None
        self._shared_transport = None
        self._ensure_storage().update_resume_metadata({})

    def needs_update(self, now: float) -> bool:
        return now >= self.next_update_at
    
    def _ensure_storage(self) -> RecordingStorage:
        """Ensure storage instance is initialized and return it."""
        if self._storage is None:
            self._storage = RecordingStorage(
                self.storage_path, 
                metadata_path=self.metadata_path,
                long_term_storage_path=self.long_term_storage_path,
                file_size_threshold=self.file_size_threshold
            )
        return self._storage

    def create_worker(self) -> ObservationWorker:
        # Check if resume data exists to decide on transport reuse
        has_resume_data = self._ensure_storage().has_resume_metadata()
        
        # For sessions with resume data, reuse the transport to reduce overhead
        if has_resume_data:
            if self._shared_transport is None:
                self._shared_transport = HTTPTransport()
            transport = self._shared_transport
        else:
            # For first updates (no resume data), create a new transport
            # This isolates the large initial response in thread memory
            transport = None

        return ObservationWorker(
            self.account,
            self.storage_path,
            self.game_id, 
            self.package, 
            self.map_cache,
            transport=transport,
            metadata_path=self.metadata_path,
            long_term_storage_path=self.long_term_storage_path,
            file_size_threshold=self.file_size_threshold
        )

    def update_package(self, other: ObservationPackage):
        self.package = other


class ObservationWorker:
    """
    Lightweight worker that performs a single update using the owning ObservationSession.
    """

    def __init__(self,
                 account: Account,
                 storage_path: Path,
                 game_id: int,
                 package: ObservationPackage = None,
                 map_cache: StaticMapCache = None,
                 transport: Optional[HTTPTransport] = None,
                 metadata_path: Optional[Path] = None,
                 long_term_storage_path: Optional[Path] = None,
                 file_size_threshold: Optional[int] = None):
        self.account = account
        self.game_id = game_id
        self.storage = RecordingStorage(
            storage_path, 
            metadata_path=metadata_path,
            long_term_storage_path=long_term_storage_path,
            file_size_threshold=file_size_threshold
        )
        self.storage.setup_logging()
        self.package: ObservationPackage = package
        self.map_cache = map_cache
        self.recording_itf = None
        self._transport = transport

    def cleanup(self):
        """Clean up resources used by this worker."""
        if self.storage:
            self.storage.teardown_logging()
            self.storage = None

    def __enter__(self):
        """Support context manager protocol."""
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        """Ensure cleanup happens when exiting context."""
        self.cleanup()
        return False


    def _on_request_response(self, response: dict):
        self.storage.update_resume_metadata(asdict(self.package))
        self.storage.save_response(response)
        del response

    def ensure_observation_package(self) -> bool:
        if self.package is not None:
            return True
        resume_metadata = self.storage.get_resume_metadata()
        if resume_metadata and "auth" in resume_metadata:
            logger.info(f"Resuming from Storage for game {self.game_id}")
            auth = AuthDetails(**resume_metadata.pop("auth"))
            resume_metadata["auth"] = auth
            self.package = ObservationPackage(**resume_metadata)
            return True

        logger.info(f"Observation package not yet created, building package for game {self.game_id}")
        self.package = self.create_observation_package()
        return True

    def create_observation_package(self) -> ObservationPackage:
        hub_itf = self.account.get_interface()
        if hub_itf is None:
            return None
        game_api = GameApi(
            session=hub_itf.api.session,
            proxy=hub_itf.api.proxy,
            game_id=self.game_id,
            auth_details=hub_itf.api.auth
        )
        game_api.load_game_site()
        return ObservationPackage(
            game_id=game_api.game_id,
            headers=dict(game_api.session.headers),
            cookies=dict(game_api.session.cookies),
            proxy=dict(game_api.session.proxies),
            auth=deepcopy(game_api.auth),
            client_version=game_api.client_version,
            game_server_address=game_api.game_server_address,
        )

    def reset_package(self):
        self.account.reset_interface()
        self.package = self.create_observation_package()

    def ensure_static_map_data(self, observation_api: ObservationApi, map_id: int) -> bool:
        if self.map_cache.is_cached(map_id):
            return True
        static_map_data = observation_api.get_static_map_data()
        self.map_cache.save(map_id, static_map_data)
        return True

    @staticmethod
    def _is_game_ended(response: Optional[dict]) -> bool:
        if not isinstance(response, dict):
            return False
        result = response.get("result")
        if not isinstance(result, dict):
            return False
        states = result.get("states", {})
        if not isinstance(states, dict):
            return False
        for state in states.values():
            if isinstance(state, dict) and state.get("gameEnded"):
                return True
        return False

    def run(self) -> bool:
        attempt = 1
        while True:
            logger.info(f"Starting update for game {self.game_id}")
            self.ensure_observation_package()

            with self._create_observation_api() as observation_api:
                try:
                    game_state = self._fetch_and_update_game_state(observation_api)
                    self._process_map_data(observation_api, game_state)

                    if self._is_game_ended(response=game_state):
                        return False

                    return True

                except AuthenticationException:
                    if not self._handle_auth_failure(attempt):
                        return False
                    attempt += 1

                except httpx.HTTPStatusError as e:
                    if e.response.status_code == 401:
                        if not self._handle_auth_failure(attempt):
                            return False
                        attempt += 1
                    elif e.response.status_code >= 500:
                        self._handle_server_error(e.response.status_code)

                except (httpx.NetworkError, httpx.ReadTimeout):
                    self._handle_network_error()

    def _create_observation_api(self) -> ObservationApi:
        # Use provided transport (reused) or create a new one (first update)
        transport = self._transport or HTTPTransport()
        api = ObservationApi(
            transport,
            headers=self.package.headers,
            cookies=self.package.cookies,
            proxy=self.package.proxy,
            auth_details=self.package.auth,
            game_id=self.game_id,
            game_server_address=self.package.game_server_address,
        )

        return api

    def _fetch_and_update_game_state(self, observation_api: ObservationApi) -> dict:
        game_state, self.package.state_ids, self.package.time_stamps = (
            observation_api.request_game_state(
                self.package.state_ids,
                self.package.time_stamps,
            )
        )

        # Update auth and connection details - only deepcopy auth as it's a complex object
        # Cookies and headers are simple dicts that can be shallow copied
        self.package.auth = deepcopy(observation_api.auth)
        self.package.cookies = dict(observation_api.client.cookies)
        self.package.headers = dict(observation_api.client.headers)
        self.package.game_server_address = observation_api.game_server_address

        self._on_request_response(game_state)
        return game_state

    def _process_map_data(self, observation_api: ObservationApi, game_state: dict) -> None:
        try:
            map_id = (
                game_state.get("result", {})
                .get("states", {})
                .get("3", {})
                .get("map", {})
                .get("mapID")
            )

            if map_id is not None:
                self.ensure_static_map_data(observation_api, int(map_id))
        except (ValueError, TypeError):
            pass  # Map data not available or invalid

    def _handle_auth_failure(self, attempt: int) -> bool:
        if attempt >= MAX_RETRIES:
            return False

        logger.warning("Authentication failed, resetting package and retrying...")
        self.reset_package()
        return True

    def _handle_server_error(self, status_code: int) -> None:
        logger.warning(
            f"GameServer returned HTTP {status_code}, "
            f"retrying in {TIME_TILL_RETRY} seconds..."
        )
        sleep(TIME_TILL_RETRY)

    def _handle_network_error(self) -> None:
        logger.warning(
            f"GameServer is not responding, "
            f"retrying in {TIME_TILL_RETRY} seconds..."
        )
        sleep(TIME_TILL_RETRY)

File: tools/server_observer/src/observation_session.cpp
#include "observation_session.hpp"
#include "observation_api.hpp"
#include <iostream>
#include <thread>
#include <chrono>
#include <utility>

static const int MAX_RETRIES = 3;
static const int TIME_TILL_RETRY = 10;


ObservationSession::ObservationSession(
    std::shared_ptr<RequestManager> manager,
    int game_id,
    std::shared_ptr<Account> account,
    std::shared_ptr<StaticMapCache> map_cache,
    std::string storage_path,
    std::string metadata_path,
    std::string long_term_storage_path,
    int file_size_threshold)
    : game_id(game_id)
      , account(std::move(account))
      , manager_(std::move(manager))
      , map_cache_(std::move(map_cache))
      , storage_path_(std::move(storage_path))
      , metadata_path_(std::move(metadata_path))
      , long_term_storage_path_(std::move(long_term_storage_path))
      , file_size_threshold_(file_size_threshold)
      , package_(), storage_(nullptr), api_(nullptr), attempt_(1) {
    next_update_at = std::chrono::system_clock::now();
}

ObservationSession::~ObservationSession() {
    if (storage_) {
        storage_->flush_metadata();  // Save metadata when session closes
        storage_->teardown_logging();
    }
}

void ObservationSession::set_proxy(const ProxyConfig& proxy_config) {
    package_.proxy = proxy_config;
}

bool ObservationSession::needs_update(std::chrono::system_clock::time_point now) const {
    return now >= next_update_at;
}

bool ObservationSession::ensure_observation_package() {
    if (package_.game_id != 0) {
        return true;
    }

    json resume_metadata = ensure_storage()->get_resume_metadata();
    if (!resume_metadata.empty() && resume_metadata.contains("auth")) {
        std::cout << "Resuming from Storage for game " << game_id << std::endl;
        package_ = ObservationPackage::from_json(resume_metadata);
        api_ = std::make_unique<ObservationApi>(
            manager_,
            package_.headers,
            package_.cookies,
            account->proxy_config,
            package_.auth,
            game_id,
            package_.game_server_address,
            package_.client_version);
        return true;
    }

    std::cout << "Observation package not yet created, building package for game "
            << game_id << std::endl;
    package_ = create_observation_package();
    return package_.game_id != 0;
}

ObservationPackage ObservationSession::create_observation_package() {
    auto hub_itf = account->get_interface();
    if (!hub_itf) {
        return {};
    }

    // Join the game as guest and load game site to get server address and auth
    std::cout << "Joining game " << game_id << " as guest to get game server data..." << std::endl;

    try {
        auto game_data = hub_itf->join_game_as_guest(game_id);

        // Convert JSON headers to std::map
        std::map<std::string, std::string> headers_map;
        if (game_data.headers.is_object()) {
            for (const auto &[key, value]: game_data.headers.items()) {
                if (value.is_string()) {
                    headers_map[key] = value.get<std::string>();
                }
            }
        }

        // Convert JSON cookies to std::map
        std::map<std::string, std::string> cookies_map;
        if (game_data.cookies.is_object()) {
            for (const auto &[key, value]: game_data.cookies.items()) {
                if (value.is_string()) {
                    cookies_map[key] = value.get<std::string>();
                }
            }
        }

        // Create an observation package with data from GameApi
        ObservationPackage pkg;
        pkg.game_id = game_id;
        pkg.headers = headers_map;
        pkg.cookies = cookies_map;
        pkg.proxy = *account->proxy_config;
        pkg.auth = game_data.auth;
        pkg.client_version = game_data.client_version;
        pkg.game_server_address = game_data.game_server_address;

        api_ = std::make_unique<ObservationApi>(
            manager_,
            pkg.headers,
            pkg.cookies,
            account->proxy_config,
            pkg.auth,
            game_id,
            pkg.game_server_address,
            pkg.client_version);
        return pkg;
    } catch (const std::exception &e) {
        std::cerr << "Failed to join game " << game_id << ": " << e.what() << std::endl;
        return {};
    }
}

void ObservationSession::reset_package() {
    if (storage_) {
        storage_->flush_metadata();  // Save metadata when package is reset
    }
    account->reset_interface();
    package_ = create_observation_package();
}

bool ObservationSession::ensure_static_map_data(ObservationApi &api, int map_id) {
    if (map_cache_->is_cached(map_id)) {
        return true;
    }

    try {
        json static_map_data = api.get_static_map_data(map_id);
        map_cache_->save(map_id, static_map_data);
        return true;
    } catch (const std::exception &e) {
        std::cerr << "Failed to get static map data: " << e.what() << std::endl;
        return false;
    }
}

// LoggingGuard implementation
ObservationSession::LoggingGuard::LoggingGuard(RecordingStorage* storage)
    : storage_(storage) {
    if (storage_) {
        storage_->setup_logging();
    }
}

ObservationSession::LoggingGuard::~LoggingGuard() noexcept {
    if (storage_) {
        storage_->teardown_logging();
    }
}

// Handle server switch scenario
ObservationResult ObservationSession::handle_server_switch(const GameServerResult& result) {
    if (!result.data.contains("newHostName")) {
        return ObservationResult::make_unknown_error("Server switch without new hostname");
    }

    std::string new_server = "https://" + result.data["newHostName"].get<std::string>();
    std::cout << "Server switch detected for game " << game_id
              << ": updating server to " << new_server << std::endl;

    // Update the package with new server address
    package_.game_server_address = new_server;

    // Recreate the ObservationApi with new server address
    api_->update_server_address(new_server);

    // Return a network error to trigger immediate retry with new server
    return ObservationResult::make_network_error(false, "Server switch to " + new_server);
}

// Convert GameServerError to ObservationResult
ObservationResult ObservationSession::handle_game_server_error(const GameServerResult& result) {
    // Handle SERVER_SWITCH specially - no attempt increment
    if (result.error_code == GameServerError::SERVER_SWITCH) {
        return handle_server_switch(result);
    }

    // Increment attempt counter for all other failures
    attempt_++;

    switch (result.error_code) {
        case GameServerError::AUTH_ERROR:
            reset_package();
            return ObservationResult::make_auth_failed(false, result.error_message);
        case GameServerError::HTTP_ERROR:
            reset_package();
            return ObservationResult::make_server_error(result.error_message);
        case GameServerError::NETWORK_ERROR:
            reset_package();
            return ObservationResult::make_network_error(false, result.error_message);
        case GameServerError::PARSE_ERROR:
            return ObservationResult::make_unknown_error(result.error_message);
        case GameServerError::UNKNOWN_ERROR:
        default:
            reset_package();
            return ObservationResult::make_unknown_error(result.error_message);
    }
}

// Update package with latest API state
void ObservationSession::update_package_from_api() {
    package_.auth = api_->get_auth();
    package_.cookies = api_->get_cookies();
    package_.headers = api_->get_headers();
    package_.game_server_address = api_->get_game_server_address();
}

// Process successful game state response
void ObservationSession::process_successful_response(GameServerResult& result) {
    // Reset attempt counter on success
    attempt_ = 0;

    // Update package with new auth and connection details
    update_package_from_api();

    // Save raw response string to storage
    ensure_storage()->update_resume_metadata(package_.to_json());
    ensure_storage()->save_response(std::move(result.raw_response));
}

asio::awaitable<ObservationResult> ObservationSession::run_update_async() {
    std::cout << "Starting update for game " << game_id
              << " (attempt " << attempt_ << ")" << std::endl;

    // RAII guard to manage storage logging lifecycle
    LoggingGuard guard(ensure_storage());

    // Ensure we have a valid observation package
    if (!ensure_observation_package()) {
        std::cerr << "Failed to create observation package" << std::endl;
        co_return ObservationResult::make_package_failed();
    }

    try {
        // Fetch game state asynchronously
        HttpResponse response = co_await api_->request_game_state_async(
            package_.state_ids,
            package_.time_stamps
        );

        // Parse and validate response
        GameServerResult result = api_->parse_and_validate_response(
            response,
            package_.state_ids,
            package_.time_stamps
        );

        // Handle errors
        if (!result.success()) {
            co_return handle_game_server_error(result);
        }

        // Process successful response
        process_successful_response(result);

        // Check if game ended
        if (result.game_ended) {
            co_return ObservationResult::make_game_ended();
        }

        co_return ObservationResult::make_success(false);
    } catch (const std::runtime_error &e) {
        co_return ObservationResult::make_unknown_error(e.what());
    }
}

void ObservationSession::set_attempt(const int attempt) {
    attempt_ = attempt;
}

int ObservationSession::get_attempt() {
    return attempt_;
}

RecordingStorage *ObservationSession::ensure_storage() {
    if (!storage_) {
        storage_ = std::make_unique<RecordingStorage>(
            storage_path_,
            metadata_path_,
            long_term_storage_path_,
            file_size_threshold_
        );
    }
    return storage_.get();
}
